### DISCUSSION TITLE
why use Window MRU position firstly?

### DISCUSSION THREAD

**Discussion by StephenRi** [ben-manes/caffeine#745] [Q&A] [answered]
I am reading the w-tinylfu algorithm implement of caffeine. I wonder why candidate = accessOrderWindowDeque.peekLast instead of peekFirst when cadidates == 0 in the evictFromMain method?
    I don't understand why use MRU position instead of LRU position firstly.

```Java
void evictFromMain(int candidates) {
    while(weightedSize() > maximum()) {
        if (candidates == 0) {
            candidate = accessOrderWindowDeque().peekLast();    // why peekLast?
        }
    }
}
```

**Comment by ben-manes** [comment, marked as answer]
This occurs for weighted caches when a large entry causes many evictions. I don’t recall if this is intentional or accidentally flipped. You can look at the commit for an explanation and suggest changes if you think that there is a mistake.

https://github.com/ben-manes/caffeine/commit/464bc1914368c47a0203517fda2151fbedaf568b

**Reply by ben-manes** [reply]
This seems intentional based on the commit message. It favors finding the excessively large entry for eviction rather than the least recently used, which could cause the cache to be flushed as it tries to free for space. It won’t happen in most scenarios and seems to be the lesser of two bad options. Of course ideas for improvements are always welcome.

**Reply by StephenRi** [reply]
Thanks for your reply. I misunderstood your intention

**Reply by StephenRi** [reply]
But，maybe there is a bad case.
When there is 3 left in Window, and 1 left in All areas. And at this time, a item weighted 2 comes. It will not be accepted, right?

Anyway, I think evicting a newly inserted item can be a little confusing. Maybe you can consider weight as a factor when admiting.

**Reply by ben-manes** [reply]
It is confusing and I’d have to review carefully to determine if it could be improved. Thankfully there are tests and a good commit message so we know the intent.

The challenge is that with concurrency we always let new entries be added or updated in the underlying map. Once that is done, we try to evict if over capacity. All evictions are done in that method, which does check if the candidate is too large and should be evicted. The alternative is to do this at the add and update tasks, but those are also complex. I agree it seems like this could be better isolated so that we aren’t hacking the MRU position and favor evictions from the LRU end. I think it requires some care to review to make those changes safely.

**Reply by ben-manes** [reply]
I tried to improve this in the v3.dev branch (https://github.com/ben-manes/caffeine/commit/a536affc3787ad85195870b9861f5932b3a5c3d8). The add/update now evicts the entry immediately in these cases, as I agree my attempt at centralizing the size eviction was misguided. The variety of exceptional cases, etc. is overwhelming so please take a look and let me know if you spot any new mistakes. I'll need to review it a few times and run some weighted traces to ensure that there are no regressions.

**Reply by ben-manes** [reply]
I ran a few weighted traces on the two versions to check for regressions.

#### Tencent Photo trace (50M - 50B)
This ranges from a 6% - 28% hit rate and is __negatively__ impacted by up to 0.3%.

#### IBM Object Store (50M - 50B)
This ranges from a 40% - 95% hit rate and is __negatively__ impacted by up to 2%.

#### MSR Cambridge - Project (50M - 500B)
This ranges from a 8% - 35% hit rate and is __positively__ impacted by up to 4%.

### Discussion
My experience is that anything within 5% is noise due to workloads differing so widely which makes it hard to draw a conclusion what is better or worse. Unfortunately, we don't have a enough weighted traces to draw a deeper analysis except that the changes appear to be within a tolerance and don't show a harmful regression. A more substantial algorithmic improvement that could improve the hit rates will come if we implement the optimizations in this [paper](https://dl.acm.org/doi/10.1145/3507920?cid=99659224047).

I think the improved clarity and the lack of a substantial difference to the hit rate makes the suggested changes worthwhile.

**Reply by chiragpujara** [reply]
Just ran into this issue during testing. Newly added entries are getting evicted in favor existing entries even though they both have same weight. IS there a timeframe when this fix will be available/released?

**Reply by ben-manes** [reply]
Can you try using the SNAPSHOT build in your test to see if it resolves your problem?

**Reply by chiragpujara** [reply]
I just tried 3.1.2-SNAPSHOT. IT has different problem or fails on different use-case. If most recent entry has highest weight than it still removes most recent entry.

**Reply by ben-manes** [reply]
The cache is not LRU. It will sample the workload and reconfigure itself over time. If you can provide an access trace we can run it through the simulator to see the overall hit rate, which is what we are trying to optimize for. In many workloads the most recent is a one-hit wonder (aka cache pollution) so we favor keeping older, more frequently used entries.
